torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+
+    if (vertex->visited) continue;
+    vertex->visited = true;
+    visited.push_back(vertex);
+
+    for (auto* next : vertex->out_edges) {
+      if (vertex != producer && next->ord == upper_bound) {
+        clearVisited(visited);
+        return true;
+      }
+      if (next->ord < upper_bound) {
+        stack.push_back(next);
+      }
+    }
+  }
+  clearVisited(visited);
+  return false;
+}
+
+
+// Performs a forward DFS.
+// Marks vertices as 'visited' and inserts them into visited.
+// Returns false immediately if upper_bound is reached.
+// Returns true otherwise.
+template <typename T>
+bool DynamicDAG<T>::dfsForward(Vertex<T>* producer, size_t upper_bound,
+    vertex_list<T>& visited) {
+  vertex_list<T> stack;
+  stack.push_back(producer);
+
+  while (!stack.empty()) {
+    auto* vertex = stack.back();
+    stack.pop_back();
+
+    if (vertex->visited) continue;
+    vertex->visited = true;
+    visited.push_back(vertex);
+
+    for (auto* next : vertex->out_edges) {
+      if (next->ord == upper_bound) {
+        return false;
+      }
+      if (next->ord < upper_bound) {
+        stack.push_back(next);
+      }
+    }
+  }
+  return true;
+}
+
+// Performs a backward DFS.
+// Marks vertices as 'visited' and inserts them into visited.
+template <typename T>
+void DynamicDAG<T>::dfsBackward(Vertex<T>* producer, size_t lower_bound,
+    vertex_list<T>& visited) {
+  vertex_list<T> stack;
+  stack.push_back(producer);
+
+  while (!stack.empty()) {
+    auto* vertex = stack.back();
+    stack.pop_back();
+
+    if (vertex->visited) continue;
+
+    vertex->visited = true;
+    visited.push_back(vertex);
+
+    for (auto* next : vertex->in_edges) {
+      if (next->ord > lower_bound) {
+        stack.push_back(next);
+      }
+    }
+  }
+}
+
+template <typename T>
+void DynamicDAG<T>::sort(vertex_list<T>& delta) {
+  struct {
+    bool operator()(Vertex<T>* a, Vertex<T>* b) const {
+      return a->ord < b->ord;
+    }
+  } custom_less;
+  std::sort(delta.begin(), delta.end(), custom_less);
+}
+
+template <typename T>
+void DynamicDAG<T>::clearVisited(vertex_list<T>& visited) {
+  for (auto* v: visited) {
+    v->visited = false;
+  }
+  visited.clear();
+}
+
+template <typename T>
+void DynamicDAG<T>::reorder(vertex_list<T>& deltaF, vertex_list<T>& deltaB) {
+  // Reorder deltaB vertices to occur before deltaF vertices.
+  sort(deltaB);
+  sort(deltaF);
+
+  vertex_list<T> tmp;
+
+  std::vector<size_t> deltaB_ords = fmap(deltaB, [](Vertex<T>* v) { return v->ord; });
+  std::vector<size_t> deltaF_ords = fmap(deltaF, [](Vertex<T>* v) { return v->ord; });
+
+  std::vector<size_t> computed_ords;
+  computed_ords.reserve(deltaB_ords.size() + deltaF_ords.size());
+  std::merge(
+    deltaB_ords.begin(), deltaB_ords.end(),
+    deltaF_ords.begin(), deltaF_ords.end(), std::back_inserter(computed_ords));
+
+  tmp.insert(tmp.end(), deltaB.begin(), deltaB.end());
+  tmp.insert(tmp.end(), deltaF.begin(), deltaF.end());
```

Comment:
Ugh 4 allocations just to do this merge sort thing (that can be done in-place)!! I guess we can optimize it in the future 😕

Reply:
plus nit: `reserve` on `tmp`